book/03-lists-and-patterns.html

```diff
@@ -634,64 +654,113 @@ <h1>Terser and Faster Patterns</h1>
-
-        <p>For simple atomic types, polymorphic compare has the
-        semantics you would expect: for floating-point numbers and
-        integers, polymorphic compare corresponds to the expected
-        numerical comparison functions. For strings, it's a
-        lexicographic comparison.</p>
-
-        <p>Sometimes, however, the type-ignoring nature of
-        polymorphic compare is a problem, particularly when you
-        have your own notion of equality and ordering that you want
-        to impose. We'll discuss this issue more, as well as some
-        of the other downsides of polymorphic compare, in <a href="13-maps-and-hashtables.html#maps-and-hash-tables" data-type="xref">Maps And Hash Tables</a>.</p>
+        <p>
+	  You might have noticed that <code>destutter</code> is
+          specialized to lists of integers. That's
+          because <code>Base</code>'s default equality operator is
+          specialized to integers, as you can see if you try to apply
+          it to values of a different type.</p>
+
+        <link rel="import" href="code/lists-and-patterns/poly.topscript" 
+	      part="1"/>
+
+	<p>
+	  OCaml also has a collection of polymorphic equality and
+	  comparison operators, which we can make available by opening
+	  the module <code>Base.Poly</code>.</p>
+	
+        <link rel="import" href="code/lists-and-patterns/poly.topscript" 
+	      part="2"/>
+
+        <p>
+	  Indeed, if we look at the type of the equality operator,
+          we'll see that it is polymorphic.</p>
+
+        <link rel="import" href="code/lists-and-patterns/poly.topscript" 
+	      part="3"/>
+
+	<p> 
+	  If we rewrite our destutter example
+	  with <code>Base.Poly</code> open, we'll see that it gets a
+	  polymorphic type, and can now be used on inputs of different
+	  types.</p>
+
+        <link rel="import" href="code/lists-and-patterns/poly.topscript" 
+	      part="4"/>
+
+        <p>
+	  OCaml comes with a whole family of polymorphic comparison
+          operators, including the standard infix
+          comparators, <code>&lt;</code>, <code>&gt;=</code>, etc., as
+          well as the function <code>compare</code> that
+          returns <code>-1</code>, <code>0</code>, or <code>1</code>
+          to flag whether the first operand is smaller than, equal to,
+          or greater than the second, respectively.</p>
+
+        <p>
+	  You might wonder how you could build functions like these
+          yourself if OCaml didn't come with them built in. It turns
+          out that you <em>can't</em> build these functions on your
+          own. OCaml's polymorphic comparison functions are built into
+          the runtime to a low level. These comparisons are
+          polymorphic on the basis of ignoring almost everything about
+          the types of the values that are being compared, paying
+          attention only to the structure of the values as they're
+          laid out in memory.</p>
```

Comment:
Perhaps forward reference to Part III and the runtime memory layout section here.